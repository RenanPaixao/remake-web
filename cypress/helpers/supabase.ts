import { createClient } from '@supabase/supabase-js'

const e2eSupabase = createClient(import.meta.env.VITE_SUPABASE_URL, import.meta.env.VITE_SUPABASE_SERVICE_ROLE_KEY, {
  auth: {
    autoRefreshToken: false,
    persistSession: false
  }
})

export { e2eSupabase }
