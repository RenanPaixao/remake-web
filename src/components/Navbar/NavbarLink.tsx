import { TheLink } from '../TheLink/TheLink.tsx'
import { PropsWithChildren } from 'react'
import { Box } from '@chakra-ui/react'
import { LinkProps } from 'react-router-dom'

export const NavbarLink = ({ children, ...rest }: PropsWithChildren<LinkProps>) => {
  return <Box>
    <TheLink {...rest}>{children}</TheLink>
  </Box>
}
