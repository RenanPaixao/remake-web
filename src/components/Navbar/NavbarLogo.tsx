import { Box, Image, ImageProps } from '@chakra-ui/react'

export const NavbarLogo = (props: ImageProps) => {
  return <Box>
    <Image {...props} fallbackSrc='https://via.placeholder.com/50' />
  </Box>
}
