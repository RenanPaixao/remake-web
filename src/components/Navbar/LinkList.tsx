import { Stack, StackItem } from '@chakra-ui/react'
import { NavbarLink } from './NavbarLink.tsx'

export const LinkList = () => {
  return <Stack w={'100%'} spacing={'8'} direction={['row']} justify={'center'} alignItems={'center'}>
    <StackItem>
      <NavbarLink to={'/'}>Home</NavbarLink>
    </StackItem>
    <StackItem>
      <NavbarLink to={'/collect-places'}>Collect Places</NavbarLink>
    </StackItem>
    <StackItem>
      <NavbarLink to={'/faq'}>FAQ</NavbarLink>
    </StackItem>
    <StackItem>
      <NavbarLink to={'/account'}>Account</NavbarLink>
    </StackItem>
  </Stack>
}
