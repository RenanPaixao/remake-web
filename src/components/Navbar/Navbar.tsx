import { Button, Flex, FlexProps } from '@chakra-ui/react'
import { NavbarLogo } from './NavbarLogo.tsx'
import { LinkList } from './LinkList.tsx'
import logo from '../../assets/logo.webp'
import { TheLink } from '../TheLink/TheLink.tsx'
import { supabase } from '../../utils/supabase.ts'
import { useState } from 'react'
import { useNavigate } from 'react-router-dom'
import { useErrorToast } from '../../hooks/toast/useErrorToast.tsx'

export const Navbar = (props: FlexProps) => {
  const errorToast = useErrorToast()
  const navigate = useNavigate()

  const [isLoading, setIsLoading] = useState(false)

  /**
   * Logs the user out of the application.
   */
  async function logout() {
    setIsLoading(true)

    const { error } = await supabase.auth.signOut()
    setIsLoading(false)

    if (error) {
      errorToast({ description: error.message })
      console.error(error)
      return
    }

    navigate('/login')
  }

  return <Flex w='100%' {...props}>
    <TheLink to={'/'}>
      <NavbarLogo h={'50px'} src={logo} />
    </TheLink>
    <LinkList />
    <Button colorScheme='red' variant='outline' w={'80px'} isLoading={isLoading} onClick={logout}>Logout</Button>
  </Flex>
}
