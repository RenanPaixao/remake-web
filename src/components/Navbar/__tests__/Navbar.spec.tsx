import { customRender } from '../../../../tests/test-utils.tsx'
import { fireEvent } from '@testing-library/dom'
import { act } from '@testing-library/react'
import { supabase } from '../../../utils/supabase.ts'
import { Navbar } from '../Navbar.tsx'

const navigateMock = vi.fn()
const supabaseSignOutMock = vi.fn()
const toastMock = vi.fn()

supabase.auth.signOut = supabaseSignOutMock
describe('Navbar', () => {
  beforeEach(() => {
    vi.resetAllMocks()

    vi.mock('react-router-dom', async () => {
      const actual: object = await vi.importActual('react-router-dom')
      return {
        ...actual,
        useNavigate: () => navigateMock
      }
    })

    vi.mock('@chakra-ui/react', async importOriginal => {
      const actual: object = await importOriginal()

      return {
        ...actual,
        useToast: () => toastMock
      }
    })
  })
  it('should render', async () => {
    const { container } = customRender(<Navbar/>)

    expect(container).toMatchSnapshot()
  })

  it('should logout successful', async () => {
    supabaseSignOutMock.mockReturnValue(Promise.resolve({
      error: null
    }))

    const { getByRole } = customRender(<Navbar/>)
    const logoutButton = getByRole('button', { name: /logout/i })

    await act(async () => {
      fireEvent.click(logoutButton)
    })

    expect(logoutButton).not.toHaveAttribute('isLoading')
    expect(navigateMock).toHaveBeenCalledWith('/login')
  })

  it('should not logout successful', async () => {
    supabaseSignOutMock.mockReturnValue(Promise.resolve({
      error: new Error('error')
    }))

    const { getByText } = customRender(<Navbar/>)
    const logoutButton = getByText(/logout/i)

    await act(async () => {
      fireEvent.click(logoutButton)
    })

    expect(logoutButton).not.toHaveAttribute('isLoading')
    expect(toastMock.mock.calls[0][0]).toEqual({
      title: 'Error',
      description: 'error',
      status: 'error',
      duration: 4000,
      isClosable: true
    })

    expect(navigateMock).not.toHaveBeenCalled()
  })
})
