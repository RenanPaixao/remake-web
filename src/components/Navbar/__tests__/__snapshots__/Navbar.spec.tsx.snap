// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Navbar > should render 1`] = `
<div>
  <div
    class="css-hboir5"
  >
    <a
      class="chakra-link css-spn4bz"
      href="/"
    >
      <div
        class="css-0"
      >
        <img
          class="chakra-image__placeholder css-16kb41"
          src="https://via.placeholder.com/50"
        />
      </div>
    </a>
    <div
      class="chakra-stack css-azg3gr"
    >
      <div
        class="chakra-stack__item css-1qy82gu"
      >
        <a
          class="chakra-link css-spn4bz"
          href="/"
        >
          <button
            class="chakra-button css-wubf5g"
            type="button"
          >
            Home
          </button>
        </a>
      </div>
      <div
        class="chakra-stack__item css-1qy82gu"
      >
        <a
          class="chakra-link css-spn4bz"
          href="/collect-places"
        >
          <button
            class="chakra-button css-wubf5g"
            type="button"
          >
            Collect Places
          </button>
        </a>
      </div>
      <div
        class="chakra-stack__item css-1qy82gu"
      >
        <a
          class="chakra-link css-spn4bz"
          href="/faq"
        >
          <button
            class="chakra-button css-wubf5g"
            type="button"
          >
            FAQ
          </button>
        </a>
      </div>
      <div
        class="chakra-stack__item css-1qy82gu"
      >
        <a
          class="chakra-link css-spn4bz"
          href="/account"
        >
          <button
            class="chakra-button css-wubf5g"
            type="button"
          >
            Account
          </button>
        </a>
      </div>
    </div>
    <button
      class="chakra-button css-obasgk"
      type="button"
    >
      Logout
    </button>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</div>
`;
